rust/user-info-fetcher/src/backend/entra.rs

```diff
@@ -0,0 +1,295 @@
+
+    #[snafu(display("unable to find user with id {user_id:?}"))]
+    UserNotFoundById {
+        source: crate::utils::http::Error,
+        user_id: String,
+    },
+
+    #[snafu(display(
+        "failed to request groups for user with username {username:?} (user_id: {user_id:?})"
+    ))]
+    RequestUserGroups {
+        source: crate::utils::http::Error,
+        username: String,
+        user_id: String,
+    },
+}
+
+impl http_error::Error for Error {
+    fn status_code(&self) -> StatusCode {
+        match self {
+            Self::AccessToken { .. } => StatusCode::BAD_GATEWAY,
+            Self::SearchForUser { .. } => StatusCode::BAD_GATEWAY,
+            Self::UserNotFoundById { .. } => StatusCode::NOT_FOUND,
+            Self::RequestUserGroups { .. } => StatusCode::BAD_GATEWAY,
+        }
+    }
+}
+
+#[derive(Deserialize)]
+struct OAuthResponse {
+    access_token: String,
+}
+
+#[derive(Clone, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct UserMetadata {
+    id: String,
+    user_principal_name: String,
+    #[serde(default)]
+    attributes: HashMap<String, serde_json::Value>,
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct GroupMembershipResponse {
+    value: Vec<GroupMembership>,
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct GroupMembership {
+    display_name: Option<String>,
+}
+
+struct EntraEndpoint {
+    hostname_token: HostName,
+    hostname_graph: HostName,
+    port: u16,
+    tenant_id: String,
+    protocol: String,
+}
+
+pub(crate) async fn get_user_info(
+    req: &UserInfoRequest,
+    http: &reqwest::Client,
+    credentials: &Credentials,
+    config: &v1alpha1::EntraBackend,
+) -> Result<UserInfo, Error> {
+    let v1alpha1::EntraBackend {
+        client_credentials_secret: _,
+        hostname_token,
+        hostname_graph,
+        port,
+        tenant_id,
+        tls,
+    } = config;
+
+    let entra_endpoint = EntraEndpoint::new(
+        hostname_token.clone(),
+        hostname_graph.clone(),
+        *port,
+        tenant_id.to_string(),
+        &TlsClientDetails { tls: tls.clone() },
+    );
+    let token_url = entra_endpoint.oauth2_token();
+
+    let authn = send_json_request::<OAuthResponse>(http.post(token_url).form(&[
+        ("client_id", credentials.client_id.as_str()),
+        ("client_secret", credentials.client_secret.as_str()),
+        ("scope", "https://graph.microsoft.com/.default"),
+        ("grant_type", "client_credentials"),
+    ]))
+    .await
+    .context(AccessTokenSnafu)?;
+
+    let user_info = match req {
+        UserInfoRequest::UserInfoRequestById(req) => {
+            let user_id = req.id.clone();
+            send_json_request::<UserMetadata>(
+                http.get(entra_endpoint.user_info(&user_id))
+                    .bearer_auth(&authn.access_token),
+            )
+            .await
+            .context(UserNotFoundByIdSnafu { user_id })?
```

Comment:
Just for the record, I think the error message is a bit misleading. It can also be a network error or permission problem.
No need to change though